### DISCUSSION TITLE
Support for multiple VectorStores

### DISCUSSION THREAD

**Discussion by pascallim** [open-webui/open-webui#4045] [Ideas]
**Description**:

Currently, the solution only supports ChromaDB as Vector DB. I want the solution to support multiple Vector DBs  without making the code too complex and specific for each Vector DB.

**Key Features**:
- Create the new code structure to be able to add easily new Vector DB
- Migrate usage of ChromaDB to this new implementation
- Add new integration with PGVector (PostgreSQL) to see how the integration for a new Vector DB will look like
- Add in the Admin Panel interface the option to choose between different Vector DBs

**Benefits**:
- The integration of new Vector DBs will be done easily
- Users won't need to migrate to ChromaDB and will be able to use their favorite Vector DB

**Technical Considerations**:
- Use of VectorStore Classes from LangChain (https://python.langchain.com/v0.2/docs/integrations/vectorstores/) that are all coming from a similar base class so we can use the same methods for each class
- One or more input attributes will change depending on the Vector DB that we want to initialize:
For Chroma DB:
![image](https://github.com/user-attachments/assets/d3d957f0-6786-40f6-a9c4-e440744844dc)

For PG Vector:
![image](https://github.com/user-attachments/assets/f7213c62-6a0e-4ca4-b45a-203021a64c87)

To handle this, we may need some utils to handle the input that will be used to initialize the LangChain classes specific to the VectorDB

**User Impact:**
- The user will be able to choose between ChromaDB and PosgreSQL as VectorDB

**Comment by james-westbrook** [comment]
I think this sounds great. Langchain is already in use in the repo, so your approach wouldn't need to add much (if anything) from a dependencies standpoint which is preferable. 

I think it would be best to leave the default ChromaDB configuration in place, and allow users to set their preferred vector DB either through environment variables if they want that present on startup, or as you mentioned, through the web UI itself. 

I'm sure other contributors/ maintainers will have thoughts, but I think this sounds like a great idea that many in the community would appreciate. Thanks for proposing this and working on it!

**Comment by justinh-rahb** [comment]
If it's all the same with Tim, this plan of action seems sound to me.

**Comment by jeremy-moffitt** [comment]
Hello,
We're interested in this feature as well, with an eye towards integrating with Milvus as the vectorDB. My team hasn't engaged with the open-webui project before but our product team has asked us to get involved explicitly in support of the vectordb choice featre. I've pointed @victorstevansuse at this thread and would love the know the best way for him to get engaged. He's got cycles available immediately. 

Thanks!

**Reply by pascallim** [reply]
> Hello, We're interested in this feature as well, with an eye towards integrating with Milvus as the vectorDB. My team hasn't engaged with the open-webui project before but our product team has asked us to get involved explicitly in support of the vectordb choice featre. I've pointed @victorstevansuse at this thread and would love the know the best way for him to get engaged. He's got cycles available immediately.
> 
> Thanks!

Hello Jeremy,

I will try to propose the design that will support easily the different vectorstores of LangChain. I will do a first round integrating with this design ChromaDB and PGVector. Victor will be able to integrate afterwards what's needed for Milvius :)

**Reply by jeremy-moffitt** [reply]
That sounds great. We'll do some prototyping in parallel and plan to integrate with your design as it becomes available.
